internal/handshake/crypto_setup_client.go

```diff
@@ -141,15 +142,21 @@ func (h *cryptoSetupClient) HandleCryptoStream(stream io.ReadWriter) error {
 		utils.Debugf("Got %s", message)
 		switch message.Tag {
 		case TagREJ:
-			err = h.handleREJMessage(message.Data)
+			if err := h.handleREJMessage(message.Data); err != nil {
+				return err
+			}
 		case TagSHLO:
-			err = h.handleSHLOMessage(message.Data)
+			params, err := h.handleSHLOMessage(message.Data)
+			if err != nil {
+				return err
+			}
+			// blocks until the session has received the parameters
```

Comment:
Nit: The API could be changed so that it guarantees this, which is currently not the case, as the chan is passed in from the outside. I'll leave it up to you.